chore: allow crater runs

bishopcheckmate opened the pull request:
## Motivation

Fixes #6924

### PR Author Checklist

#### Check before marking the PR as ready for review:
  - [x] Will the PR name make sense to users?
  - [x] Does the PR have a priority label?
  - [x] Have you added or updated tests?
  - [x] Is the documentation up to date?

##### For significant changes:
  - [x] Is there a summary in the CHANGELOG?
  - [x] Can these changes be split into multiple PRs?

### Specifications

This change adds versions to all dev-dependencies so that they end up being packed together with crates on crates.io thus allowing crater runs.

It also removes the circular dependency between `tower-batch-control` and `zebra-consensus` by copying the Ed25519 verifier code into batch-control tests. Also updates the verifier tests in consensus to their state from tower-batch-control.

There are still unadressed points in #6924:

> Make sure the tests that are on by default:
> - run reasonably fast
> - don't use too much network
> - (check crater docs for any other requirements)

It's hard for me to define reasonably fast and too much networking. I can try to measure those on my PC if you want.
Regarding crater docs for any other requirements, I couldn't find any guidelines on that.

### Testing

This should be tested using `cargo publish`, however it's not currently possible due to git dependencies added in #8136.

### Reviewer Checklist

Check before approving the PR:
  - [ ] Does the PR scope match the ticket?
  - [ ] Are there enough tests to make sure it works? Do the tests cover the PR motivation?
  - [ ] Are all the PR blockers dealt with?
        PR blockers can be dealt with in new tickets or PRs.

_And check the PR Author checklist is complete._

bishopcheckmate committed:
chore(dev-deps): always specify the version for dev-dependencies

bishopcheckmate committed:
test(tower-batch-control): remove zebra-consensus dev dependency

Copies the Ed25519Verifier code directly to tower-batch-control tests

bishopcheckmate committed:
test(zebra-consensus): update ed25519 verifier tests with ones from batch-control

arya2 commented on tower-batch-control/tests/ed25519.rs:
```diff
@@ -1,18 +1,175 @@
+    }
+}
+
+impl Service<BatchControl<Item>> for Verifier {
```
Good idea defining the service here for testing. 

Optional: Since `zebra-consensus` has the same coverage now, I think it would be good to simplify this test service to remove the dependencies on `zebra-chain` and `metrics`. Let me know if you'd like me to open a suggestion/follow-up PR for this change.

bishopcheckmate replied:
Thanks,  I'll check

bishopcheckmate replied:
I've removed the metrics. with zebra-chain tho, would you like me to also replicate batch verifier in the batch control tests or something else?

arya2 replied:
> with zebra-chain tho, would you like me to also replicate batch verifier in the batch control tests or something else?

I was thinking of changing `Item` to be a `Result`, so the test verifier looks for the first item that's an error and only tests the `Batch` service.

This PR looks good to merge if you'd like to skip it.

bishopcheckmate replied:
I've just pushed the commit which uses just `ed25519` and not the whole `zebra-chain`

bishopcheckmate replied:
> This PR looks good to merge if you'd like to skip it.

> only tests the Batch service

I think I see the idea and I'd like to do that but I need to get the better understanding of the service to come up with some nice simple examples. I am not sure I'll have time for it before the weekend so we can merge it and I'll do a follow up if that's okay

arya2 replied:
> I think I see the idea and I'd like to do that but I need to get the better understanding of the service to come up with some nice simple examples. I am not sure I'll have time for it before the weekend so we can merge it and I'll do a follow up if that's okay

That sounds great, this PR has already gone beyond what was needed to close the issue (thank you).

I can open the follow up PR for your review if you'd like, I think it'll be easier to understand the service and the suggestion that way.

bishopcheckmate replied:
If you don't mind I like the harder path, it gives more project understanding

bishopcheckmate committed:
test(zebra-consensus): restore previous timeout values

The timeouts copied from tower-batch-control seems too low
and there must have been a reasoning why were they introduced

bishopcheckmate committed:
Merge branch 'main' into chore/allow-crater-runs

bishopcheckmate committed:
chore: update dev-deps versions to beta.33

arya2 reviewed (dismissed):
This looks great, thank you!

I left an optional suggestion for removing the dependency on zebra-chain and metrics.

> Make sure the tests that are on by default:
> 
> run reasonably fast
> don't use too much network
> (check crater docs for any other requirements)

I didn't find any guidelines either, I'm not sure if there are any constraints in the code, or if it's manually observed and configured on a per-project basis here: https://github.com/rust-lang/crater/blob/master/config.toml

bishopcheckmate committed:
chore(tower-batch-control): remove dev-dependency on metrics

bishopcheckmate committed:
chore(tower-batch-control): remove zebra-chain dev-dependency

arya2 reviewed (dismissed):
This looks excellent, thank you!

arya2 committed:
Merge branch 'main' into chore/allow-crater-runs

oxarbitrage committed:
Merge branch 'main' into chore/allow-crater-runs

arya2 commented:
The CI failure seems unrelated to the changes, we should try to fix that or merge this manually.

arya2 committed:
Merge branch 'main' into chore/allow-crater-runs

oxarbitrage commented:
@bishopcheckmate just FYI: your code is approved to be part of the zebra main branch but we are having CI issues with external pull requests. Right now, @gustavovalverde is working on fixing this issue. We will keep you posted, thanks for your contribution. I hope this problem do not discourage you to submit more code.

bishopcheckmate commented:
hey, no problem at all. I'll definitely keep contributing here and there, just been busy recently

arya2 committed:
Merge branch 'main' into chore/allow-crater-runs

arya2 commented on zebra-scan/Cargo.toml:
```diff
@@ -55,7 +55,15 @@ zcash_client_backend = "0.10.0-rc.1"
+  "clock",
+  "std",
+  "serde",
+] }
```
I resolved the merge conflict incorrectly here.

```suggestion
```

arya2 committed:
Update zebra-scan/Cargo.toml

arya2 committed:
Merge branch 'main' into chore/allow-crater-runs

oxarbitrage committed:
Merge branch 'main' into chore/allow-crater-runs

oxarbitrage commented on tower-batch-control/Cargo.toml:
```diff
@@ -43,8 +46,7 @@ rand = "0.8.5"
 tokio = { version = "1.36.0", features = ["full", "tracing", "test-util"] }
 tokio-test = "0.4.3"
-tower-fallback = { path = "../tower-fallback/" }
+tower-fallback = { path = "../tower-fallback/", version = "0.2.41-beta.8" }
```
```suggestion
tower-fallback = { path = "../tower-fallback/", version = "0.2.41-beta.10" }
```

oxarbitrage commented on tower-batch-control/Cargo.toml:
```diff
@@ -43,8 +46,7 @@ rand = "0.8.5"
 
-zebra-consensus = { path = "../zebra-consensus/" }
-zebra-test = { path = "../zebra-test/" }
+zebra-test = { path = "../zebra-test/", version = "1.0.0-beta.33" }
```
```suggestion
zebra-test = { path = "../zebra-test/", version = "1.0.0-beta.34" }
```

oxarbitrage commented on tower-fallback/Cargo.toml:
```diff
@@ -24,4 +24,4 @@ tracing = "0.1.39"
 tokio = { version = "1.36.0", features = ["full", "tracing", "test-util"] }
 
-zebra-test = { path = "../zebra-test/" }
+zebra-test = { path = "../zebra-test/", version = "1.0.0-beta.33" }
```
```suggestion
zebra-test = { path = "../zebra-test/", version = "1.0.0-beta.34" }
```

oxarbitrage commented on zebra-chain/Cargo.toml:
```diff
@@ -168,7 +168,7 @@ rand_chacha = "0.3.1"
 tokio = { version = "1.36.0", features = ["full", "tracing", "test-util"] }
 
-zebra-test = { path = "../zebra-test/" }
+zebra-test = { path = "../zebra-test/", version = "1.0.0-beta.33" }
```
```suggestion
zebra-test = { path = "../zebra-test/", version = "1.0.0-beta.34" }
```

oxarbitrage commented on zebra-consensus/Cargo.toml:
```diff
@@ -94,6 +94,6 @@ tokio = { version = "1.36.0", features = ["full", "tracing", "test-util"] }
-zebra-test = { path = "../zebra-test/" }
+zebra-state = { path = "../zebra-state", version = "1.0.0-beta.33", features = ["proptest-impl"] }
+zebra-chain = { path = "../zebra-chain", version = "1.0.0-beta.33", features = ["proptest-impl"] }
+zebra-test = { path = "../zebra-test/", version = "1.0.0-beta.33" }
```
```suggestion
zebra-state = { path = "../zebra-state", version = "1.0.0-beta.34", features = ["proptest-impl"] }
zebra-chain = { path = "../zebra-chain", version = "1.0.0-beta.34", features = ["proptest-impl"] }
zebra-test = { path = "../zebra-test/", version = "1.0.0-beta.34" }
```

oxarbitrage commented on zebra-rpc/Cargo.toml:
```diff
@@ -87,9 +87,9 @@ proptest = "1.4.0"
+zebra-chain = { path = "../zebra-chain", version = "1.0.0-beta.33", features = ["proptest-impl"] }
+zebra-consensus = { path = "../zebra-consensus", version = "1.0.0-beta.33", features = ["proptest-impl"] }
+zebra-network = { path = "../zebra-network", version = "1.0.0-beta.33", features = ["proptest-impl"] }
+zebra-state = { path = "../zebra-state", version = "1.0.0-beta.33", features = ["proptest-impl"] }
```
```suggestion
zebra-chain = { path = "../zebra-chain", version = "1.0.0-beta.34", features = ["proptest-impl"] }
zebra-consensus = { path = "../zebra-consensus", version = "1.0.0-beta.34", features = ["proptest-impl"] }
zebra-network = { path = "../zebra-network", version = "1.0.0-beta.34", features = ["proptest-impl"] }
zebra-state = { path = "../zebra-state", version = "1.0.0-beta.34", features = ["proptest-impl"] }
```

oxarbitrage commented on zebra-rpc/Cargo.toml:
```diff
@@ -87,9 +87,9 @@ proptest = "1.4.0"
+zebra-state = { path = "../zebra-state", version = "1.0.0-beta.33", features = ["proptest-impl"] }
 
-zebra-test = { path = "../zebra-test" }
+zebra-test = { path = "../zebra-test", version = "1.0.0-beta.33" }
```
```suggestion
zebra-test = { path = "../zebra-test", version = "1.0.0-beta.34" }
```

oxarbitrage commented on zebra-scan/Cargo.toml:
```diff
@@ -54,6 +54,7 @@ futures = "0.3.30"
 zcash_client_backend = "0.10.0-rc.1"
 zcash_primitives = "0.13.0-rc.1"
 
+
```
```suggestion
```

oxarbitrage commented on zebra-script/Cargo.toml:
```diff
@@ -25,4 +25,4 @@ displaydoc = "0.2.4"
 hex = "0.4.3"
 lazy_static = "1.4.0"
-zebra-test = { path = "../zebra-test" }
+zebra-test = { path = "../zebra-test", version = "1.0.0-beta.33" }
```
```suggestion
zebra-test = { path = "../zebra-test", version = "1.0.0-beta.34" }
```

oxarbitrage commented on zebra-state/Cargo.toml:
```diff
@@ -107,5 +107,5 @@ jubjub = "0.10.0"
-zebra-chain = { path = "../zebra-chain", features = ["proptest-impl"] }
-zebra-test = { path = "../zebra-test/" }
+zebra-chain = { path = "../zebra-chain", version = "1.0.0-beta.33", features = ["proptest-impl"] }
+zebra-test = { path = "../zebra-test/", version = "1.0.0-beta.33" }
```
```suggestion
zebra-chain = { path = "../zebra-chain", version = "1.0.0-beta.34", features = ["proptest-impl"] }
zebra-test = { path = "../zebra-test/", version = "1.0.0-beta.34" }
```

oxarbitrage commented on zebrad/Cargo.toml:
```diff
@@ -280,13 +280,13 @@ proptest-derive = "0.4.0"
+zebra-consensus = { path = "../zebra-consensus", version = "1.0.0-beta.33", features = ["proptest-impl"] }
+zebra-network = { path = "../zebra-network", version = "1.0.0-beta.33", features = ["proptest-impl"] }
+zebra-scan = { path = "../zebra-scan", version = "0.1.0-alpha.1", features = ["proptest-impl"] }
+zebra-state = { path = "../zebra-state", version = "1.0.0-beta.33", features = ["proptest-impl"] }
```
```suggestion
zebra-chain = { path = "../zebra-chain", version = "1.0.0-beta.34", features = ["proptest-impl"] }
zebra-consensus = { path = "../zebra-consensus", version = "1.0.0-beta.34", features = ["proptest-impl"] }
zebra-network = { path = "../zebra-network", version = "1.0.0-beta.34", features = ["proptest-impl"] }
zebra-scan = { path = "../zebra-scan", version = "0.1.0-alpha.3", features = ["proptest-impl"] }
zebra-state = { path = "../zebra-state", version = "1.0.0-beta.34", features = ["proptest-impl"] }
```

oxarbitrage commented on zebrad/Cargo.toml:
```diff
@@ -280,13 +280,13 @@ proptest-derive = "0.4.0"
+zebra-state = { path = "../zebra-state", version = "1.0.0-beta.33", features = ["proptest-impl"] }
 
-zebra-node-services = { path = "../zebra-node-services", features = ["rpc-client"] }
+zebra-node-services = { path = "../zebra-node-services", version = "1.0.0-beta.33", features = ["rpc-client"] }
```
```suggestion
zebra-node-services = { path = "../zebra-node-services", version = "1.0.0-beta.34", features = ["rpc-client"] }
```

oxarbitrage commented on zebrad/Cargo.toml:
```diff
@@ -280,13 +280,13 @@ proptest-derive = "0.4.0"
+zebra-node-services = { path = "../zebra-node-services", version = "1.0.0-beta.33", features = ["rpc-client"] }
 
 zebra-test = { path = "../zebra-test" }
 zebra-grpc = { path = "../zebra-grpc" }
```
```suggestion
zebra-test = { path = "../zebra-test" version = "1.0.0-beta.34" }
zebra-grpc = { path = "../zebra-grpc" version = "0.1.0-alpha.1" }
```

oxarbitrage commented on zebrad/Cargo.toml:
```diff
@@ -300,4 +300,4 @@ zebra-grpc = { path = "../zebra-grpc" }
 # https://github.com/rust-lang/cargo/issues/9096
 # zebra-utils { path = "../zebra-utils", artifact = "bin:zebra-checkpoints" }
-zebra-utils = { path = "../zebra-utils" }
+zebra-utils = { path = "../zebra-utils", version = "1.0.0-beta.33" }
```
```suggestion
zebra-utils = { path = "../zebra-utils", version = "1.0.0-beta.34" }
```

oxarbitrage reviewed (commented):
Bumping all the versions to the ones in our current release.

oxarbitrage committed:
bump all versions to match current release

oxarbitrage commented on zebrad/Cargo.toml:
```diff
@@ -280,16 +280,16 @@ proptest-derive = "0.4.0"
-zebra-test = { path = "../zebra-test" }
-zebra-grpc = { path = "../zebra-grpc" }
+zebra-test = { path = "../zebra-test" version = "1.0.0-beta.34" }
+zebra-grpc = { path = "../zebra-grpc" version = "0.1.0-alpha.1" }
```
missed commas, sorry.

```suggestion
zebra-test = { path = "../zebra-test", version = "1.0.0-beta.34" }
zebra-grpc = { path = "../zebra-grpc", version = "0.1.0-alpha.1" }
```

oxarbitrage committed:
fix missed commas in version bumps

mpguerra committed:
Merge branch 'main' into chore/allow-crater-runs
